File: src/main/java/org/distril/beengine/util/BedrockResourceLoader.java
package org.distril.beengine.util;

import com.nukkitx.nbt.NbtMap;
import com.nukkitx.nbt.NbtUtils;
import org.distril.beengine.Bootstrap;

import java.io.IOException;

public class BedrockResourceLoader {

	public static final NbtMap BIOME_DEFINITIONS;
	public static final NbtMap ENTITY_IDENTIFIERS;

	static {
		try (var reader = NbtUtils.createNetworkReader(Bootstrap.getResource("data/biome_definitions.dat"))) {
			BIOME_DEFINITIONS = (NbtMap) reader.readTag();
		} catch (IOException exception) {
			throw new RuntimeException(exception);
		}

		try (var reader = NbtUtils.createNetworkReader(Bootstrap.getResource("data/entity_identifiers.dat"))) {
			ENTITY_IDENTIFIERS = (NbtMap) reader.readTag();
		} catch (IOException exception) {
			throw new RuntimeException(exception);
		}
	}
}

File: src/main/kotlin/org/distril/beengine/util/BedrockResourceLoader.kt
package org.distril.beengine.util

import com.nukkitx.nbt.NbtMap
import com.nukkitx.nbt.NbtUtils
import com.nukkitx.protocol.bedrock.packet.AvailableEntityIdentifiersPacket
import com.nukkitx.protocol.bedrock.packet.BiomeDefinitionListPacket
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

object BedrockResourceLoader {

	val biomeDefinitionListPacket = BiomeDefinitionListPacket()

	val availableEntityIdentifiersPacket = AvailableEntityIdentifiersPacket()

	suspend fun init() = coroutineScope {
		launch {
			NbtUtils.createNetworkReader(Utils.getResource("data/biome_definitions.dat")).use {
				biomeDefinitionListPacket.definitions = it.readTag() as NbtMap
			}
		}

		launch {
			NbtUtils.createNetworkReader(Utils.getResource("data/entity_identifiers.dat")).use {
				availableEntityIdentifiersPacket.identifiers = it.readTag() as NbtMap
			}
		}
	}
}
